### DISCUSSION TITLE
How to build an universal distribution for different Python version

### DISCUSSION THREAD

**Discussion by inlann** [pypa/setuptools#3587] [Q&A]
Hello:

I developed a Python extension module in C++ and generated a wrap cxx file for it using [swig](https://www.swig.org/).

I'm trying to build a distribution with the setuptools. I'm using the Python 3.10 and my `setup.py` is:

```python
#!/usr/bin/python3
#-*- coding:utf-8 -*-


from setuptools import setup, Extension


shared_ads_module = Extension(
    name='_SharedADS',
    sources=[
        'sharedADS_wrap.cxx',  # generated by swig
        'src/ads/ADSLexer.cpp',
        'src/ads/ADSParser.cpp',
        'src/ads/Decl.cpp',
        'src/ads/Serializer.cpp',
        'src/ads/SharedADS.cpp',
    ],
    include_dirs=['include'],
    language='c++',
    extra_compile_args=['-std=c++17']
)

setup(
    name='SharedADS',
    version='1.0',
    author='Terapines',
    description='ads parser module',
    ext_modules=[shared_ads_module],
    py_modules=['SharedADS']
)
```

The command `python3 setup.py bdist build_ext --inplace` generates a `SharedADS.py` and a `_SharedADS.cp310-win_amd64.pyd`. But the `_SharedADS.cp310-win_amd64.pyd` does not work on other machines that do not use Python 3.10 which means I need to rebuild it for different Python version.

Is there any way to build an universal distribution for different Python version?

**Comment by abravalheri** [comment]
Hi @inlann, I am afraid that packages with binary extensions are naturally non-universal.
Compiled files are coupled with OS, system architecture, version of the processor instruction set, etc...

You can try to have a look on [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) for ways of building different versions of your extensions for multiple systems.

**Reply by inlann** [reply]
Thanks for your reply 😄 
I'm sorry I didn't describe the problem clearly. I only want to use it on Windows, but the compiled result is coupled with the Python version. So, I'm trying to find a way to build an universal distribution for different Python version.
